include/rocksdb/file_system.h

```diff
@@ -1340,6 +1340,15 @@ class FileSystemWrapper : public FileSystem {
 
 class FSSequentialFileWrapper : public FSSequentialFile {
  public:
+  // Creates a FileWrapper around the input File object and takes
+  // ownership of the object
+  explicit FSSequentialFileWrapper(std::unique_ptr<FSSequentialFile>&& t)
```

Comment:
We already have inheritance and a virtual destructor, so why not make a derived class of `FSSequentialFileWrapper` called `FSSequentialFileOwnerWrapper` that takes ownership of the target and deallocates on destruction? I don't like using extra fields to encode in dynamic state of fields something that could be encoded in static and dynamic types. And I don't think it'll be much more code to create the new derived classes.

Reply:
I can do that but am not sure how much you are gaining by it.  The class basically only has a constructor/destructor and everything else is the same, correct?

- What argument does the constructor take -- a raw pointer or a unique one?
- Does the Owner class store the wrapped pointer (as a unique) or just access the base?  Is the wrapped pointer explicitly destroyed?

Reply:
The owner version would take unique_ptr.

If `FSSequentialFileOwnerWrapper : public FSSequentialFileWrapper` then I think simplest to destroy target_ in destructor, though you could add another copy of the pointer as unique_ptr if you prefer.

There's also a possible solution where the kind of pointer is templated, but that's probably overkill.

Reply:
> The owner version would take unique_ptr.
> 
> If `FSSequentialFileOwnerWrapper : public FSSequentialFileWrapper` then I think simplest to destroy target_ in destructor, though you could add another copy of the pointer as unique_ptr if you prefer.

Would destroying `target_` in `FSSequentialFileOwnerWrapper`'s destructor leave its `FSSequentialFileWrapper` subobject with a dangling pointer when its destructor runs?

Reply:
There are a couple ways of doing this, and I am not sure which we prefer.  I am open to ideas:
1.  The "Owner" class can have a std::unique_ptr.  When the Owner is destroyed, the unique ptr will be cleaned and target_ will become invalid.  In this case, the wrapped file is implicitly deleted
2. The "Owner" class can delete target_ in its destructor (explicitly deleted).

In both cases, we could reset target_ if desired so it would point to nullptr instead of invalid (freed) memory.

Preferences?